File: go/src/methods/loadstrike_scenario/create_method_reference.go
package loadstrike_scenario

        import (
        	"time"

        	loadstrike "loadstrike.com/sdk/go"
        )

        const createRunnerKey = "runner_dummy_orders_reference"

type CreateMethodReference struct{}

type createTempConfigPaths struct {
	ConfigPath string
	InfraPath  string
}

type createOrdersReportingSink struct{}

func newCreateOrdersReportingSink() createOrdersReportingSink {
	return createOrdersReportingSink{}
}
func (createOrdersReportingSink) SinkName() string {
	return "orders-sample-sink"
}
func (createOrdersReportingSink) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) SaveRealtimeStats([]loadstrike.LoadStrikeScenarioStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) SaveRealtimeMetrics(loadstrike.LoadStrikeMetricStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) SaveRunResult(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersReportingSink) Dispose() {}

type createOrdersRuntimePolicy struct{}

func newCreateOrdersRuntimePolicy() createOrdersRuntimePolicy {
	return createOrdersRuntimePolicy{}
}
func (createOrdersRuntimePolicy) ShouldRunScenario(string) loadstrike.LoadStrikeBoolTask {
	return loadstrike.TaskFromBool(true)
}
func (createOrdersRuntimePolicy) BeforeScenario(string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersRuntimePolicy) AfterScenario(string, loadstrike.LoadStrikeScenarioRuntime) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersRuntimePolicy) BeforeStep(string, string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersRuntimePolicy) AfterStep(string, string, loadstrike.LoadStrikeReply) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

type createOrdersWorkerPlugin struct{}

func newCreateOrdersWorkerPlugin() createOrdersWorkerPlugin {
	return createOrdersWorkerPlugin{}
}
func (createOrdersWorkerPlugin) PluginName() string {
	return "orders-sample-plugin"
}
func (createOrdersWorkerPlugin) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersWorkerPlugin) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersWorkerPlugin) GetData(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikePluginData] {
	return loadstrike.TaskFromResult(loadstrike.CreatePluginData("orders-sample-plugin"))
}
func (createOrdersWorkerPlugin) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (createOrdersWorkerPlugin) Dispose() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

func createPerformOrderGetReply() loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeResponse.Ok("200", int64(128), "ok", loadstrike.TimeSpan(3*time.Millisecond))
}

func createExecuteOrderGet(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeStep.Run("get-order", context, func(loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
		return createPerformOrderGetReply()
	})
}

func createExecuteOrderGetAsync(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikeReply] {
	return loadstrike.TaskFromResult(createExecuteOrderGet(context))
}

func createBaselineScenario(names ...string) loadstrike.LoadStrikeScenario {
	name := "orders.get-by-id"
	if len(names) > 0 && names[0] != "" {
		name = names[0]
	}
	return loadstrike.CreateScenario(name, createExecuteOrderGet).
		WithLoadSimulations(loadstrike.LoadStrikeSimulation.IterationsForConstant(1, 1)).
		WithoutWarmUp()
}

func createBaseRunner() loadstrike.LoadStrikeRunner {
	return loadstrike.Create().
		AddScenario(createBaselineScenario()).
		WithRunnerKey(createRunnerKey).
		WithTestSuite("orders-reference").
		WithTestName("orders-get-by-id").
		WithoutReports()
}

func createBaseContext() loadstrike.LoadStrikeContext {
	return createBaseRunner().BuildContext()
}

func createHttpSource() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-source",
		Mode:          "Produce",
		TrackingField: "header:X-Correlation-Id",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                   "https://orders.example.test/api/orders",
			Method:                "GET",
			TrackingPayloadSource: "Request",
			ResponseSource:        "ResponseBody",
		},
	}
}

func createHttpDestination() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-destination",
		Mode:          "Consume",
		TrackingField: "json:$.trackingId",
		GatherByField: "json:$.tenantId",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                      "https://orders.example.test/api/order-events",
			Method:                   "GET",
			ResponseSource:           "ResponseBody",
			ConsumeJSONArrayResponse: true,
			ConsumeArrayPath:         "$.items",
		},
	}
}

func createTrackingConfiguration() *loadstrike.TrackingConfigurationSpec {
	return &loadstrike.TrackingConfigurationSpec{
		Source:                      createHttpSource(),
		Destination:                 createHttpDestination(),
		RunMode:                     "GenerateAndCorrelate",
		CorrelationTimeoutSeconds:   30,
		TimeoutSweepIntervalSeconds: 1,
		TimeoutBatchSize:            200,
		TimeoutCountsAsFailure:      true,
		MetricPrefix:                "orders_tracking",
		ExecuteOriginalScenarioRun:  false,
	}
}

func createTrackedContext() loadstrike.LoadStrikeContext {
	return loadstrike.Create().
		AddScenario(createBaselineScenario("orders.tracked").WithCrossPlatformTracking(createTrackingConfiguration())).
		WithRunnerKey(createRunnerKey).
		WithoutReports().
		BuildContext()
}

func createBuildTrackingPayload() loadstrike.TrackingPayload {
	builder := loadstrike.TrackingPayloadBuilder{
		Headers:            map[string]string{"X-Correlation-Id": "ord-1"},
		ContentType:        "application/json",
		MessagePayloadType: "json",
	}
	builder.SetBody(map[string]any{"trackingId": "ord-1"})
	return builder.Build()
}

func createRunResult() loadstrike.LoadStrikeRunResult {
	var result loadstrike.LoadStrikeRunResult
	return result
}

func createScenarioStats() loadstrike.LoadStrikeScenarioStats {
	var stats loadstrike.LoadStrikeScenarioStats
	return stats
}

func createWriteTempConfigFiles() createTempConfigPaths {
	return createTempConfigPaths{
		ConfigPath: "method-reference.loadstrike.config.json",
		InfraPath:  "method-reference.loadstrike.infra.json",
	}
}

        // Create the default GET /orders/{id} scenario.
func (reference CreateMethodReference) CreateBaselineScenarioExample() any {
    return loadstrike.CreateScenario("orders.get-by-id", createExecuteOrderGet)
}

// Create the baseline scenario and immediately add a load simulation.
func (reference CreateMethodReference) CreateScenarioAndAddSimulationExample() any {
    return loadstrike.CreateScenario("orders.get-by-id", createExecuteOrderGet).WithLoadSimulations(loadstrike.LoadStrikeSimulation.IterationsForConstant(1, 3))
}

File: java/src/main/java/com/loadstrike/samplereference/methods/loadstrike_runner/CreateMethodReference.java
package com.loadstrike.samplereference.methods.loadstrike_runner;

      import com.loadstrike.runtime.CorrelationStoreConfiguration;
      import com.loadstrike.runtime.CrossPlatformScenarioConfigurator;
      import com.loadstrike.runtime.CrossPlatformTrackingConfiguration;
      import com.loadstrike.runtime.HttpEndpointDefinition;
      import com.loadstrike.runtime.LoadStrikeCorrelation.TrackingFieldSelector;
      import com.loadstrike.runtime.LoadStrikeCorrelation.TrackingPayload;
      import com.loadstrike.runtime.LoadStrikeCorrelation.TrackingPayloadBuilder;
      import com.loadstrike.runtime.LoadStrikeRuntime.*;
      import com.loadstrike.runtime.LoadStrikeTransports;
      import java.time.Duration;
      import java.util.Map;
      import java.util.concurrent.CompletableFuture;
      import java.util.concurrent.CompletionStage;

      public final class CreateMethodReference {
        private CreateMethodReference() {}

        private static final String RUNNER_KEY = "runner_dummy_orders_reference";

private static LoadStrikeScenario baselineScenario() {
  return baselineScenario("orders.get-by-id");
}

private static LoadStrikeScenario baselineScenario(String name) {
  return LoadStrikeScenario.create(name, context -> executeOrderGet(context))
      .withLoadSimulations(LoadStrikeSimulation.iterationsForConstant(1, 1))
      .withoutWarmUp();
}

private static LoadStrikeReply<?> executeOrderGet(LoadStrikeScenarioContext context) {
  return LoadStrikeStep.run("get-order", context, () -> createOrderReply()).asReply();
}

private static CompletionStage<LoadStrikeReply<?>> executeOrderGetAsyncStage(
    LoadStrikeScenarioContext context) {
  return CompletableFuture.completedFuture(executeOrderGet(context));
}

private static LoadStrikeReply<Map<String, Object>> createOrderReply() {
  return LoadStrikeResponse.ok(Map.of("orderId", "ORD-10001"), "200", 128, "ok", 3.2d);
}

private static LoadStrikeRunner baseRunner() {
  return LoadStrikeRunner.create()
      .addScenario(baselineScenario())
      .withRunnerKey(RUNNER_KEY)
      .withTestSuite("orders-reference")
      .withTestName("orders-get-by-id")
      .withoutReports();
}

private static LoadStrikeContext baseContext() {
  return baseRunner().buildContext();
}

private static LoadStrikeContext trackedContext() {
  return LoadStrikeRunner.create()
      .addScenario(
          CrossPlatformScenarioConfigurator.Configure(
              baselineScenario("orders.tracked"), trackingConfiguration()))
      .withRunnerKey(RUNNER_KEY)
      .withoutReports()
      .buildContext();
}

private static HttpEndpointDefinition httpSource() {
  var endpoint = new HttpEndpointDefinition();
  endpoint.name = "orders-http-source";
  endpoint.mode = LoadStrikeTransports.TrafficEndpointMode.Produce;
  endpoint.trackingField = TrackingFieldSelector.parse("header:X-Correlation-Id");
  endpoint.url = "https://orders.example.test/api/orders";
  endpoint.method = "GET";
  endpoint.responseSource = "ResponseBody";
  return endpoint;
}

private static HttpEndpointDefinition httpDestination() {
  var endpoint = new HttpEndpointDefinition();
  endpoint.name = "orders-http-destination";
  endpoint.mode = LoadStrikeTransports.TrafficEndpointMode.Consume;
  endpoint.trackingField = TrackingFieldSelector.parse("json:$.trackingId");
  endpoint.gatherByField = TrackingFieldSelector.parse("json:$.tenantId");
  endpoint.url = "https://orders.example.test/api/order-events";
  endpoint.method = "GET";
  endpoint.responseSource = "ResponseBody";
  endpoint.consumeJsonArrayResponse = true;
  endpoint.consumeArrayPath = "$.items";
  return endpoint;
}

private static CrossPlatformTrackingConfiguration trackingConfiguration() {
  var tracking = new CrossPlatformTrackingConfiguration();
  tracking.source = httpSource();
  tracking.destination = httpDestination();
  tracking.runMode = LoadStrikeTransports.TrackingRunMode.GenerateAndCorrelate;
  tracking.correlationTimeout = Duration.ofSeconds(30);
  tracking.timeoutSweepInterval = Duration.ofSeconds(1);
  tracking.timeoutBatchSize = 200;
  tracking.timeoutCountsAsFailure = true;
  tracking.metricPrefix = "orders_tracking";
  tracking.executeOriginalScenarioRun = false;
  tracking.correlationStore = CorrelationStoreConfiguration.inMemory();
  return tracking;
}

private static TrackingPayload buildTrackingPayload() {
  var builder = new TrackingPayloadBuilder();
  builder.setBody("{\"trackingId\":\"ord-1\"}");
  return builder.build();
}

private static LoadStrikeScenarioContext createStepContext() {
  return null;
}

private static LoadStrikeRunResult runResult() {
  return null;
}

private static LoadStrikeScenarioStats scenarioStats() {
  return null;
}

private static TempConfigPaths writeTempConfigFiles() {
  return new TempConfigPaths("method-reference.loadstrike.config.json", "method-reference.loadstrike.infra.json");
}

private record TempConfigPaths(String ConfigPath, String InfraPath) {}

private static final class OrdersReportingSink implements LoadStrikeReportingSink {
  @Override
  public String sinkName() {
    return "orders-sample-sink";
  }

  @Override
  public void init(LoadStrikeBaseContext context, Map<String, Object> infraConfig) {}

  @Override
  public void start(LoadStrikeSessionStartInfo sessionInfo) {}

  @Override
  public void saveRealtimeStats(LoadStrikeScenarioStats[] scenarioStats) {}

  @Override
  public void saveRealtimeMetrics(LoadStrikeMetricStats metrics) {}

  @Override
  public void saveRunResult(LoadStrikeRunResult result) {}

  @Override
  public void stop() {}

  @Override
  public void dispose() {}
}

private static final class OrdersRuntimePolicy implements LoadStrikeRuntimePolicy {
  @Override
  public boolean shouldRunScenario(String scenarioName) {
    return true;
  }

  @Override
  public void beforeScenario(String scenarioName) {}

  @Override
  public void afterScenario(String scenarioName, LoadStrikeScenarioRuntime stats) {}

  @Override
  public void beforeStep(String scenarioName, String stepName) {}

  @Override
  public void afterStep(String scenarioName, String stepName, LoadStrikeReply<?> reply) {}
}

private static final class OrdersWorkerPlugin implements LoadStrikeWorkerPlugin {
  @Override
  public String pluginName() {
    return "orders-sample-plugin";
  }

  @Override
  public void init() {}

  @Override
  public void start(Map<String, Object> session) {}

  @Override
  public LoadStrikePluginData getData(LoadStrikeRunResult result) {
    return LoadStrikePluginData.create(pluginName());
  }
}

        /** Create an empty runner before scenarios are added. */
public static Object CreateEmptyRunnerExample() {
    return LoadStrikeRunner.create();
}

/** Create a runner and immediately attach the baseline GET scenario. */
public static Object CreateRunnerAndAddScenarioExample() {
    return LoadStrikeRunner.create().addScenario(baselineScenario());
}
      }
